File: src/main/java/org/elasticsearch/plugin/readonlyrest/acl/blocks/rules/impl/ActionsRule.java
package org.elasticsearch.plugin.readonlyrest.acl.blocks.rules.impl;

import org.elasticsearch.common.logging.ESLogger;
import org.elasticsearch.common.logging.Loggers;
import org.elasticsearch.common.settings.Settings;
import org.elasticsearch.plugin.readonlyrest.acl.RequestContext;
import org.elasticsearch.plugin.readonlyrest.acl.blocks.rules.MatcherWithWildcards;
import org.elasticsearch.plugin.readonlyrest.acl.blocks.rules.Rule;
import org.elasticsearch.plugin.readonlyrest.acl.blocks.rules.RuleExitResult;
import org.elasticsearch.plugin.readonlyrest.acl.blocks.rules.RuleNotConfiguredException;

/**
 * Created by sscarduzio on 14/02/2016.
 */
public class ActionsRule extends Rule {

  private final static ESLogger logger = Loggers.getLogger(ActionsRule.class);

  protected MatcherWithWildcards m;

  public ActionsRule(Settings s) throws RuleNotConfiguredException {
    super(s);
    m = new MatcherWithWildcards(s, KEY);
  }


  @Override
  public RuleExitResult match(RequestContext rc) {
    if(m.match(new String[]{rc.getAction()})){
      return MATCH;
    }
    logger.debug("This request uses the action'" + rc.getAction() + "' and none of them is on the list.");
    return NO_MATCH;
  }
}

File: corex/src/main/scala/tech/beshu/ror/acl/blocks/rules/ActionsRule.scala
package tech.beshu.ror.acl.blocks.rules

import cats.data.NonEmptySet
import cats.implicits._
import org.apache.logging.log4j.scala.Logging
import monix.eval.Task
import tech.beshu.ror.acl.blocks.BlockContext
import tech.beshu.ror.acl.blocks.rules.ActionsRule.Settings
import tech.beshu.ror.acl.blocks.rules.Rule.{RegularRule, RuleResult}
import tech.beshu.ror.acl.request.RequestContext
import tech.beshu.ror.acl.aDomain.Action
import tech.beshu.ror.acl.show.logs._
import tech.beshu.ror.utils.MatcherWithWildcards

import scala.collection.JavaConverters._

class ActionsRule(val settings: Settings)
  extends RegularRule with Logging {

  override val name: Rule.Name = ActionsRule.name

  private val matcher = new MatcherWithWildcards(settings.actions.map(_.value).toSortedSet.asJava)

  override def check(requestContext: RequestContext,
                     blockContext: BlockContext): Task[RuleResult] = Task {
    if (matcher.`match`(requestContext.action.value)) {
      RuleResult.Fulfilled(blockContext)
    } else {
      logger.debug(s"This request uses the action '${requestContext.action.show}' and none of them is on the list.")
      RuleResult.Rejected
    }
  }
}

object ActionsRule {

  val name = Rule.Name("actions")

  final case class Settings(actions: NonEmptySet[Action])

}
